### ISSUE TITLE
elliptic-curve: `JwkEcKey` is too strict in what it accepts

### ISSUE THREAD

**Issue by Erik1000** [RustCrypto/traits#1016]
As per [RFC7517 section 4](https://datatracker.ietf.org/doc/html/rfc7517#section-4) (fourth paragraph)
> Additional members can be present in the JWK; if not understood by
   implementations encountering them, they MUST be ignored.

But currently, if you try to parse this key for example (using the `p256` crate; key created by [step-cli](https://github.com/smallstep/cli)):
```json
{
  "use": "enc",
  "kty": "EC",
  "kid": "UYa89vgc4u_lpcbbmDQfYJQRAUD4AED8H8FUNjk5KyQ",
  "crv": "P-256",
  "alg": "ECDH-ES",
  "x": "hFc6OfbgRsYFOWyhGbWH0sS5DZBjJLGABJvPttVZfA4",
  "y": "tnXB8ks0-AZJKOgbMWJrE5Jm3nTFy0UiqQugmx9jku4"
}
```

It will fail, because it does not recognize the `use` parameter (it doesn't recognize some other parameters either, but that is the first parameter it sees).
In fact, you have to reduce it to the following parameters (so only the ones from [`Field`](https://docs.rs/elliptic-curve/latest/src/elliptic_curve/jwk.rs.html#370-376) are present)

```json
{
  "kty": "EC",
  "crv": "P-256",
  "x": "hFc6OfbgRsYFOWyhGbWH0sS5DZBjJLGABJvPttVZfA4",
  "y": "tnXB8ks0-AZJKOgbMWJrE5Jm3nTFy0UiqQugmx9jku4"
}
```

There's a registry for all key parameters at [IANA](https://www.iana.org/assignments/jose/jose.xhtml#web-key-parameters) but note that this list  1) may grow with new entries and 2) implementations are allowed to define custom parameters not in the registry.

My proposed solution would be to ignore unknown parameters just like the spec dictates it.

**Comment by Erik1000** [comment]
I think the whole parsing from Json Web Keys and the abstraction over different EC key types does not belong into this crate, tbh.
You can have a look how we implemented JsonWebKeys in `jose` https://docs.rs/jose/latest/jose/struct.JsonWebKey.html

**Comment by ubamrein** [comment]
Yeah I was wondering, how one then would use e.g. the `p256` crate. Because until now I loved the ease of access to get a verification key from a JWK.

**Comment by ubamrein** [comment]
I agreee that processing of the whole JWK RFC should be out of scope. But constructing a Public/Private key from a JWK is something I really appreciate when working with the rust-crypto crates. 

So maybe the parsing could just made lenient? Like ignore anything but `x`, `y`, `d` , `kty` and `crv` or so?

**Comment by Erik1000** [comment]
I think the RFC specifies parameters that MUST BE understood and I agree that other parameters (OPTIONAL ones) should be simply ignored if not understood.

**Comment by ubamrein** [comment]
Sure, but thats 30 lines of code vs. 1 line of code ;)

**Comment by tarcieri** [comment]
The current implementation avoids `serde_derive` but I think it's fine to switch to the custom derive which would ignore unknown fields automatically

**Comment by ubamrein** [comment]
> [@ubamrein](https://github.com/ubamrein) you could open a PR to relax deserialization of JwkEcKey and therefore close this issue

There is already a PR that's why I asked what the state is

**Comment by tarcieri** [comment]
Aah yes #1547, which stalled with me asking if we should retire the implementation entirely 😉

**Comment by ubamrein** [comment]
> Aah yes [#1547](https://github.com/RustCrypto/traits/pull/1547), which stalled with me asking if we should retire the implementation entirely 😉

exactly :D

**Comment by tarcieri** [comment]
There's also https://github.com/RustCrypto/JOSE/tree/master/jose-jwk although offhand I'm not sure if it provides equivalent functionality

**Comment by Erik1000** [comment]
> There's also https://github.com/RustCrypto/JOSE/tree/master/jose-jwk although offhand I'm not sure if it provides equivalent functionality

It does only handle (de)serialization it seems. I do not think using it is worth it, because it basically deserializes into primitive types like `String` and `Vec<Bytes>`.
We should keep it simple and any parsing of JWK parameters is out of scope IMO (or do we want to parse X.509 certificate chains? I dont think so)

**Comment by ubamrein** [comment]
> > There's also https://github.com/RustCrypto/JOSE/tree/master/jose-jwk although offhand I'm not sure if it provides equivalent functionality
> 
> It does only handle (de)serialization it seems. I do not think using it is worth it, because it basically deserializes into primitive types like `String` and `Vec<Bytes>`. We should keep it simple and any parsing of JWK parameters is out of scope IMO (or do we want to parse X.509 certificate chains? I dont think so)

Yes I agree. As I said, really liked the one liner to quickly verify e.g. JWTs :)  

I'll try to come up with a PR, to be sure:

- Use custom derive
- Add `use` as a argument (for the sake of best practice)
- Ignore unknown properties 

Does that sound reasonable?

**Comment by Erik1000** [comment]
If you do so, remember the `use` parameter is optional
